### PR TITLE
Apply peformance tuning to new sandboxes also

### PR DESCRIPTION
**Pull request by evol262**
relates to #35082, moby/libnetwork#2491

Previously, values for expire_quiescent_template, conn_reuse_mode,
and expire_nodest_conn were set only system-wide. Also apply them
for new lb_* and ingress_sbox sandboxes, so they are appropriately
propagated

Signed-off-by: Ryan Barry rbarry@mirantis.com

Signed-off-by: Ryan Barry <rbarry@mirantis.com>

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: libnetwork/osl/namespace_linux.go
Comment (samuelkarp) [inline libnetwork/osl/namespace_linux.go:684]: This is failing to compile:

```
libnetwork/osl/namespace_linux.go:682:25: undefined: loadBalancerConfig
libnetwork/osl/namespace_linux.go:683:3: duplicate case SandboxTypeLoadBalancer (value 1) in switch
previous case at libnetwork/osl/namespace_linux.go:681:3
```

```suggestion
		case SandboxTypeLoadBalancer, SandboxTypeIngress:
			kernel.ApplyOSTweaks(map[string]*kernel.OSValue{
```

I'm also curious: why move the values inline here instead of retaining the `loadBalancerConfig` package-level variable?

**Review by samuelkarp — DISMISSED** [review]

**Comment by thaJeztah** [comment]
Looks like this replaces https://github.com/moby/moby/pull/42897 (looks like there was a bad rebase on that one?)

I see there's a typo in the commit message (`peformance` instead of `performance`); if you're updating, perhaps you could fix that one as well (in case we need to search in git history).

> I'm also curious: why move the values inline here instead of retaining the loadBalancerConfig package-level variable?

I think that was my suggestion in https://github.com/moby/moby/pull/42897/commits/f014fa20df7f27fdfbe8ef8c1e8771e2ef9cb49f#r718371177, where I initially thought both were setting the same options. The variable was only used in a single place, so I suggested to either move them closer to where they're used, or just to inline them (to reduce the cognitive overload of finding "where" they're used, and the reverse ("what options are in `loadBalancerConfig` and `ingressConfig` ?)

That said, the earlier version did set two different options in two places; see https://github.com/moby/moby/pull/42897/commits/f014fa20df7f27fdfbe8ef8c1e8771e2ef9cb49f, so I'm a bit confused now why this PR is different from the earlier iteration; I think this PR is missing some of the changes from https://github.com/moby/moby/pull/42897/commits/f014fa20df7f27fdfbe8ef8c1e8771e2ef9cb49f / https://github.com/moby/moby/pull/42897

**Comment by evol262** [comment]
Yep -- the rebase did clobber the old PR when trying to sign off one of the commits. 

The change to `libnetwork/controller.go` also got missed when I pulled it over. Added now. @thaJeztah is exactly right about the inlining -- it was a suggestion from the previous PR

**Review by thaJeztah — CHANGES_REQUESTED** [review]
changes look ok, but looks like one change is in the wrong commit; can you fix that?

**Review by corhere — APPROVED** [review]

**Comment by thaJeztah** [comment]
(kicked CI)


@corhere is https://github.com/moby/moby/pull/43146#discussion_r875300377 ok with you, or do you still prefer to have it addressed?

File: libnetwork/osl/namespace_linux.go
Comment (corhere) [reply libnetwork/osl/namespace_linux.go:681]: Setting a sysctl may be fast, but it isn't free: it requires opening, writing to, and closing a `procfs` file and therefore costs three syscalls. Three syscalls is still three context switches into the kernel and back, even when setting the sysctl value is ultimately a no-op. Adding a `CheckFn` would not reduce the number of context switches as the file still would need to be opened and read from.

I'm not following; how does redundantly setting the sysctl add safety? What's being made safer?

To be clear, I have no problem with blindly setting the sysctls, or setting them redundantly on the same network namespace over time. It's just that setting the syscalls N times _in a tight loop_ when there's a simple way to hoist them out of the loop seems unnecessarily wasteful.

**Comment by corhere** [comment]
@thaJeztah I would prefer to have it addressed, but I wouldn't block merging on it

**Review by thaJeztah — APPROVED** [review]
LGTM (kicked CI once more just to be sure, see my comment higher up)

**Comment by thaJeztah** [comment]
Oh; I was wrong; looks like it is a known flaky (but on specific architectures); https://github.com/moby/moby/issues/41561

For posterity; failure was;

```
=== RUN   TestNetworkLocalhostTCPNat
    nat_test.go:55: assertion failed: hi yall (msg string) !=  (string)
--- FAIL: TestNetworkLocalhostTCPNat (0.99s)
```

anyway, let's see if CI goes green and get this in

**Comment by thaJeztah** [comment]
Bringing this one in; CI is green now 👍

**Comment by xinfengliu** [comment]
I found another issue. The `sysctl` settings are not applied after the 1st time starting docker daemon since OS boot. Then after `systemctl restart docker`, the settings are applied.
```shell
<right after booting OS>
docker@ub2204-1:~$ sudo nsenter --net=/var/run/docker/netns/lb_wgyv3nagr sysctl net.ipv4.vs.expire_nodest_conn
net.ipv4.vs.expire_nodest_conn = 0
docker@ub2204-1:~$ sudo sysctl net.ipv4.vs.expire_nodest_conn
net.ipv4.vs.expire_nodest_conn = 0

<restart docker daemon>
docker@ub2204-1:~$ sudo systemctl restart docker

docker@ub2204-1:~$ sudo nsenter --net=/var/run/docker/netns/lb_wgyv3nagr sysctl net.ipv4.vs.expire_nodest_conn
net.ipv4.vs.expire_nodest_conn = 1
docker@ub2204-1:~$ sudo sysctl net.ipv4.vs.expire_nodest_conn
net.ipv4.vs.expire_nodest_conn = 1
```
This means the settings would get lost after the user reboots the system and they probably would not notice that.

**Comment by thaJeztah** [comment]
@xinfengliu nice find! Looks like there's some other code-path that should also apply it.

Could you perhaps open a ticket for that? Just to be sure it's not forgotten (easy to miss comments on merged pull requests); contributions welcome as well of course!

**Comment by xinfengliu** [comment]
@thaJeztah 

After further checking, I find it is because `ip_vs` module is not loaded by default in my system, starting docker triggers loading `ip_vs`, but when docker tries to setting the ipvs sysctl parameters, the `ip_vs` module has not been loaded yet.

```shell
May 22 13:22:53 ub2204-1 dockerd[1053]: time="2022-05-22T13:22:53.350537245Z" level=error msg="error reading the kernel parameter net.ipv4.vs.conn_reuse_mode" error="open /proc/sys/net/ipv4/vs/conn_reuse_mode: no such file or directory"
May 22 13:22:53 ub2204-1 dockerd[1053]: time="2022-05-22T13:22:53.350561161Z" level=error msg="error reading the kernel parameter net.ipv4.vs.expire_nodest_conn" error="open /proc/sys/net/ipv4/vs/expire_nodest_conn: no such file or directory"
May 22 13:22:53 ub2204-1 dockerd[1053]: time="2022-05-22T13:22:53.350567369Z" level=error msg="error reading the kernel parameter net.ipv4.vs.expire_quiescent_template" error="open /proc/sys/net/ipv4/vs/expire_quiescent_template: no such file or directory"
May 22 13:22:53 ub2204-1 dockerd[1053]: time="2022-05-22T13:22:53.350577244Z" level=error msg="error reading the kernel parameter net.ipv4.vs.conn_reuse_mode" error="open /proc/sys/net/ipv4/vs/conn_reuse_mode: no such file or directory"
```
Not sure if we should create an moby issue for it (addressing it in moby code).

**Pull request merged**
